### DISCUSSION TITLE
[question] ZodType with ZodObject props

### DISCUSSION THREAD

**Discussion by domosedov** [colinhacks/zod#1886] [Q&A] [answered]
Hello, how create zodtype with object props from external types?

```ts
// Generated type from API
type User = {
 id: number
 name: string
}

const UserSchema = z.ZodType<User> = z.object({
  id: z.number(),
  name: z.string()
})

UserSchema.pick  // not working
// Property 'pick' does not exist on type 'ZodType<User, ZodTypeDef, User>'.
```

**Comment by suhaotian** [comment]
```ts

type User = {
  id: number;
  name: string;
};

const UserSchema = z.object({
  id: z.number(),
  name: z.string(),
});

type USER = z.infer<typeof UserSchema>;


```

![2022-10-15 at 11 09 PM](https://user-images.githubusercontent.com/6512574/195993770-41a12274-e137-40a1-bea3-a5a65710fc4a.png)

**Comment by nicolapalavecino** [comment]
Maybe you are looking to construct a validation object (`z.object(...)`) with a type. I believe that's not possible because of the compile time/run time limitations. I suppose you are looking something like this:

```typescript
// Generated type from API
type User = {
  id: number;
  name: string;
};

const UserSchema: z.ZodObject<any, any, any, User> = z.object({
  id: z.number(),
  name: z.string(),
});

UserSchema.pick; // now it works
```

**Comment by domosedov** [comment, marked as answer]
Modern solution TS 4.9+
```ts
type User = {
    id: number
    name: string
    age: number
}

const UserSchema = z.object({
    id: z.number(),
    name: z.string(),
    age: z.number()
}) satisfies z.ZodType<User>
```

**Reply by victorvat** [reply]
Great solution!
And if you have optional fields in your type definition and don't want to forget to include them into Your Zod schema:
```
type User = {
    id: number
    name?: string | null
    age?: number | null
}

const UserSchema = z.object({
    id: z.number(),
    name: z.string().optional().nullable(),
    age: z.number().optional().nullable()
}) satisfies z.ZodType<Required<User>>
```

**Comment by cyrilchapon** [comment]
@domosedov thanks for this.

While it helps a lot, it fails at :
- providing autocompletion inside z.object shape
- underlining correctly errors inside z.object shape
- detecting extra-properties on the schema

**Comment by rob-gordon** [comment]
This solution is really helpful, but I wish it forwarded JSDoc information. That may be way out of scope, if so I apologize. Here are a couple images illustrating what I mean (check out where the cursor is)

Property description

![Screenshot 2023-04-01 at 12 08 42 PM](https://user-images.githubusercontent.com/159949/229302079-d9882b66-c04b-466e-a0ac-c09b0d633d2e.png)

Zod parse with satisfies type... no description :(

![Screenshot 2023-04-01 at 12 08 51 PM](https://user-images.githubusercontent.com/159949/229302097-28c68c10-6970-4143-a317-3e6e5ac6f080.png)

**Comment by LeulAria** [comment]
i just end up using this implementation since i only want the keys for intellecence while adding validation so i get types from db scheam or gql generate or other then based on that i validate so it's fine for my usecase.

```ts
export type InferZodMap<T extends abstract new (...args: any) => any> = {
    [k in keyof Partial<InstanceType<T>>]?: unknown;
};

// then use it in zod

type User {
    email: string;
}

const UserInsertValidation = z.object({
  email: z.string(),
} satisfies InferZodMap<typeof User>);
```


this one doesn't seem to work form me
```ts
const UserInsertValidation = z.object({
  email: z.string(),
}) satisfies ZodType<User>;
```
